figtree.go

```diff
@@ -691,375 +707,1044 @@ func (m *Merger) mustIgnore(name string) bool {
+		return false
 	}
+	if option := toOption(dst); option != nil {
+		return true
+	}
+	if _, ok := dst.Interface().(yaml.Node); ok {
+		return true
+	}
+	return false
 }
 
-func (m *Merger) mergeArrays(ov, nv reflect.Value) reflect.Value {
+func (m *Merger) mergeArrays(dst reflect.Value, src mergeSource, overwrite bool) (reflect.Value, bool, error) {
 	var cp reflect.Value
-	switch ov.Type().Kind() {
+	switch dst.Type().Kind() {
 	case reflect.Slice:
-		cp = reflect.MakeSlice(ov.Type(), ov.Len(), ov.Len())
-		reflect.Copy(cp, ov)
+		if overwrite {
+			// overwriting so just make a new slice
+			cp = reflect.MakeSlice(dst.Type(), 0, 0)
+		} else {
+			cp = reflect.MakeSlice(dst.Type(), dst.Len(), dst.Len())
+			reflect.Copy(cp, dst)
+		}
 	case reflect.Array:
 		// arrays are copied, not passed by reference, so we dont need to copy
-		cp = ov
+		cp = dst
+	}
+
+	if !src.isList() {
+		reflectedSrc, coord, err := src.reflect()
+		if err != nil {
+			return reflect.Value{}, false, err
+		}
+		if !reflectedSrc.IsValid() {
+			// if this is a nil interface data then
+			// we don't care that we cant assign it to a
+			// list, it is a no-op anyway.
+			return cp, false, nil
+		}
+		return reflect.Value{}, false, errors.WithStack(
+			notAssignableError{
+				srcType:        reflectedSrc.Type(),
+				dstType:        dst.Type(),
+				sourceLocation: NewSource(m.sourceFile, WithLocation(coord)),
+			},
+		)
 	}
 	var zero interface{}
-Outer:
-	for ni := 0; ni < nv.Len(); ni++ {
-		niv := nv.Index(ni)
-
-		n := niv
-		if n.CanAddr() {
-			if nOption, ok := n.Addr().Interface().(option); ok {
-				if !nOption.IsDefined() {
-					continue
+	changed := false
+	err := src.foreach(func(ix int, item mergeSource) error {
+		reflected, _, err := item.reflect()
+		if err != nil {
+			return err
+		}
+		if dst.Kind() == reflect.Array {
+			if dst.Len() <= ix {
+				// truncate arrays, we cannot append
+				return nil
+			}
+			dstElem := dst.Index(ix)
+			if isDefault(dstElem) || dstElem.IsZero() || overwrite {
+				ok, err := m.assignValue(dstElem, item, assignOptions{
+					Overwrite: overwrite,
+				})
+				if err != nil {
+					return err
 				}
-				n = reflect.ValueOf(nOption.GetValue())
+				changed = changed || ok
 			}
+			return nil
 		}
 
-		if reflect.DeepEqual(n.Interface(), zero) {
-			continue
+		// if src or dst's are options we need to compare the
+		// values to determine if we need to skip inserting this
+		// element
+		compareValue := reflected
+		if nOption := toOption(reflected); nOption != nil {
+			if !nOption.IsDefined() {
+				return nil
+			}
+			compareValue = reflect.ValueOf(nOption.GetValue())
 		}
 
-		for oi := 0; oi < cp.Len(); oi++ {
-			o := cp.Index(oi)
-			if o.CanAddr() {
-				if oOption, ok := o.Addr().Interface().(option); ok {
-					o = reflect.ValueOf(oOption.GetValue())
+		if !compareValue.IsValid() || reflect.DeepEqual(compareValue.Interface(), zero) {
+			return nil
+		}
+
+		for i := 0; i < cp.Len(); i++ {
```

Comment:
I guess this isn't new but the `O(n^2)` behavior feels problematic. I'm not sure if there's any good solution. In a codebase I used to work on I had code for hashing arbitrary reflected values to a SHA256 digest, but it was kind of a pain.

I don't like the "skip duplicates" behavior to begin with... I think it makes a lot of assumptions about the semantics of arrays, and it would generally be better for the figtree user to remove duplicates where appropriate, using efficient code. But I suspect we can't change the behavior at this point.

Reply:
yeah, not sure I would do it this way again, but I am not sure how we would back out of this now.  We can optimize this a bit if needed by caching some of the intermediates, but it has never shown up in our performance benchmarking, so not sure it is worth bothering with now.